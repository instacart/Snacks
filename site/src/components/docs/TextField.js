import React, {Fragment} from 'react'
import Playground from '../Playground'

export default () => (

TextField Example:

<Playground>
{`
<div style={{width: '335px'}}>
  <div style={{marginBottom: '10px'}}>
    <TextField
      style={{marginRight: '5px'}}
      name="first_name"
      type="text"
      floatingLabelText="First Name"
      halfWidth
      required
    />
    <TextField
      style={{marginLeft: '5px'}}
      name="last_name"
      type="text"
      floatingLabelText="Last Name"
      halfWidth
      required
    />
  </div>
  <div style={{marginBottom: '10px'}}>
    <TextField
      name="email"
      type="email"
      floatingLabelText="Email"
      hintText="Enter your email address"
      validations={{isEmail: null, isLength: {min: 3, max: 15}}}
      validationErrorText="Please enter a valid email"
      fullWidth
      required
    />
  </div>

  <div>
    <TextField
      name="password"
      type="password"
      floatingLabelText="Password"
      hintText="Enter a secure password"
      validations={{isLength: {min: 6}}}
      validationErrorText="Password must be at least 6 characters"
      fullWidth
      required
    />
  </div>
</div>
`}
</Playground>

)
