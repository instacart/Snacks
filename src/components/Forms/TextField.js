import React             from 'react'
import PropTypes         from 'prop-types'
import Radium            from 'radium'
import { FormComponent } from 'ic-formable'
import { colors }        from '../../styles'
import ValidationError   from './ValidationError'
import FloatingLabel     from './FloatingLabel'
import TextFieldHint     from './TextFieldHint'
import ServerError       from './ServerError'
import HelperText        from './HelperText'

const styles = {
  wrapper: {
    cursor: 'auto',
    display: 'inline-block',
    position: 'relative',
    width: '343px',
  },
  inputContainer: {
    borderRadius: '4px',
    position: 'relative'
  },
  input: {
    backgroundColor: '#FFF',
    border: `solid 1px ${colors.GRAY_74}`,
    borderRadius: '4px',
    boxSizing: 'border-box',
    color: colors.GRAY_20,
    fontSize: '16px',
    height: '56px',
    margin: '0',
    padding: '25px 8px 8px 8px',
    outline: 'none',
    position: 'relative',
    width: '100%',
    WebkitOpacity: 1,
    WebkitTapHighlightColor: 'rgba(0,0,0,0)'
  },
  highlight: {
    border: `1px solid ${colors.GREEN_500}`
  },
  inputDisabled: {
    border: `1px dashed ${colors.GRAY_74}`,
    backgroundColor: colors.GRAY_93,
    color: colors.GRAY_46,
    cursor: 'not-allowed'
  },
  inputError: {
    border: `1px solid ${colors.RED_700}`,
    backgroundColor: '#FDE6EB'
  },
  fullWidth: {
    width: '100%'
  },
  halfWidth: {
    width: '162px'
  }
}

@FormComponent
@Radium
class TextField extends React.Component {
  static propTypes = {
    /** Name of the field */
    name               : PropTypes.string.isRequired,
    /** HTML autocomplete attribute */
    autoComplete       : PropTypes.string,
    /** DefaultValue for non controlled component */
    defaultValue       : PropTypes.any,
    /** Disable the text field */
    disabled           : PropTypes.bool,
    /** Text of label that will animate when TextField is focused */
    floatingLabelText  : PropTypes.string,
    /** Sets width to 100% */
    fullWidth          : PropTypes.bool,
    /** Sets width to 162px */
    halfWidth          : PropTypes.bool,
    /** Helper text will show up in bottom right corner below TextField */
    helperText         : PropTypes.string,
    /** Hint text will show up when input is focused and there is no value */
    hintText           : PropTypes.string,
    /** Style for hint text */
    hintTextStyle      : PropTypes.object,
    /** Uniq id for input */
    id                 : PropTypes.string,
    /** Style for input */
    inputStyle         : PropTypes.object,
    /** Set by FormComponent by default.   */
    isValid            : PropTypes.bool,
    /** onFocus callback */
    onFocus            : PropTypes.func,
    /** onChange callback */
    onChange           : PropTypes.func,
    /** onBlur callback */
    onBlur             : PropTypes.func,
    /** onKeyDown callback */
    onKeyDown          : PropTypes.func,
    /** Mark the field as required.  */
    required           : PropTypes.bool,
    /** Error from server to show ServerError message */
    serverError        : PropTypes.string,
    /** Wrapper styles */
    style              : PropTypes.object,
    /** Input type ie. 'text', 'email', password, etc..  */
    type               : PropTypes.string,
    /** Text to show for validation error  */
    validationErrorText: PropTypes.string,
    /** Value will make TextField a controlled component  */
    value              : PropTypes.string,
  }

  static defaultProps = {
    autoComplete: 'on',
    disabled    : false,
    type        : 'text',
    defaultValue: null
  }

  state = {
    hasValue: this.props.defaultValue !== null || Boolean(this.props.value)
  }

  componentWillMount() {
    const { id, name } = this.props

    // uniqueId is needed label htmlFor properties
    this.uniqueId = id || `${name}-${Math.floor(Math.random() * 0xFFFF)}`.replace(/[^A-Za-z0-9-]/gi, '')
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.disabled && !this.props.disabled) {
      this.setState({ isFocused: false })
    }

    if (!this.state.hasValue && nextProps.value) {
      this.setState({hasValue: true})
    }
  }

  handleInputChange = (e) => {
    const { onChange } = this.props
    const { hasValue } = this.state
    const value        = e.target.value
    // Limit setState call to only when hasValue changes
    if (value && !hasValue) {
      this.setState({hasValue: true})
    } else if (!value && hasValue) {
      this.setState({hasValue: false})
    }

    onChange && onChange(e, value)
  }

  handleInputFocus = (e) => {
    this.setState({isFocused: true})
    this.props.onFocus && this.props.onFocus(e)
  }

  handleInputBlur = (e) => {
    this.setState({isFocused: false})
    this.props.onBlur && this.props.onBlur(e)
  }

  render() {
    const {
      floatingLabelText,
      defaultValue,
      disabled,
      fullWidth,
      halfWidth,
      hintText,
      inputStyle,
      isValid,
      name,
      required,
      serverError,
      type,
      validationErrorText,
      style,
      value,
      helperText,
      onKeyDown,
      autoComplete
    } = this.props

    const {
      hasValue,
      isFocused
    } = this.state

    const inputId = this.uniqueId
    const showHintText = hintText && !hasValue && isFocused
    const hasError = (!disabled && (!isValid && isValid !== undefined)) || !!serverError

    return (
      <div
        style={[
          styles.wrapper,
          fullWidth && styles.fullWidth,
          halfWidth && styles.halfWidth,
          style
        ]}
      >

        {serverError && !disabled && !isValid &&
          <ServerError
            text={serverError}
          />
        }

        <div style={styles.inputContainer}>
          <FloatingLabel
            text={floatingLabelText}
            float={isFocused || hasValue}
            disabled={disabled}
            isActive={isFocused}
            hasError={hasError}
            htmlFor={inputId}
          />

          { hintText &&
            <TextFieldHint
              text={hintText}
              show={showHintText}
              disabled={disabled}
            />
          }

          <input
            value={value}
            id={inputId}
            ref="input"
            defaultValue={value !== undefined ? undefined : defaultValue}
            disabled={disabled}
            name={name}
            type={type}
            aria-required={required}
            aria-invalid={hasError}
            aria-describedby={hasError ? `error_${inputId}` : null}
            style={[
              styles.input,
              inputStyle,
              disabled && styles.inputDisabled,
              !disabled && hasError && styles.inputError,
              isFocused && !hasError && styles.highlight
            ]}
            onBlur={this.handleInputBlur}
            onChange={this.handleInputChange}
            onFocus={this.handleInputFocus}
            onKeyDown={onKeyDown && onKeyDown()}
            autoComplete={autoComplete}
          />
        </div>

        <ValidationError
          text={validationErrorText}
          show={!disabled && !isValid && !serverError}
          inputId={inputId}
        />

        <HelperText helperText={helperText} />
      </div>
    )
  }
}

export default TextField
