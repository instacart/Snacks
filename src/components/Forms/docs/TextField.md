TextField Example:

```js
<div>
  <TextField
    name="test"
    type="email"
    floatingLabelText="Email"
    hintText="Enter your email address"
  />
</div>
```
