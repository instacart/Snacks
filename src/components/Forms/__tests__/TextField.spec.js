import React from 'react'
import renderer from 'react-test-renderer'
import { StyleRoot } from 'radium'
import { mount } from 'enzyme'
import { spy } from 'sinon'
import TextField from '../TextField'

const defaultTextField = (
  <StyleRoot>
    <div>
      <TextField
        id="test_id"
        name="test"
        type="email"
        floatingLabelText="Email"
        hintText="Enter your email address"
      />
    </div>
  </StyleRoot>
)

it('renders TextField correctly', () => {
  const tree = renderer.create(defaultTextField).toJSON()
  expect(tree).toMatchSnapshot()
})

it('fires the onFocus prop', () => {
  const onFocus = spy()
  const wrapper = mount(
    <StyleRoot>
      <div>
        <TextField
          id="test_id"
          name="test"
          type="email"
          floatingLabelText="Email"
          hintText="Enter your email address"
          onFocus={onFocus}
        />
      </div>
    </StyleRoot>
  )

  wrapper.find('input').simulate('focus')

  expect(onFocus.calledOnce).toBe(true)
})

it('fires the triggerFocus method', () => {
  const wrapper = mount(defaultTextField)

  wrapper
    .find('TextField')
    .first()
    .instance()
    .triggerFocus()
  setTimeout(() => {
    expect(wrapper.children().matchesElement(document.activeElement)).toEqual(
      true,
      'The input was not focused'
    )
  }, 10)
})

it('fires the onBlur prop', () => {
  const onBlur = spy()
  const wrapper = mount(
    <StyleRoot>
      <div>
        <TextField
          id="test_id"
          name="test"
          type="email"
          floatingLabelText="Email"
          hintText="Enter your email address"
          onBlur={onBlur}
        />
      </div>
    </StyleRoot>
  )

  wrapper.find('input').simulate('blur')

  expect(onBlur.calledOnce).toBe(true)
})

it('fires the onChange prop', () => {
  const onChange = spy()
  const wrapper = mount(
    <StyleRoot>
      <div>
        <TextField
          id="test_id"
          name="test"
          type="email"
          floatingLabelText="Email"
          hintText="Enter your email address"
          onChange={onChange}
        />
      </div>
    </StyleRoot>
  )

  wrapper.find('input').simulate('change', { target: { value: 'My new value' } })

  expect(onChange.calledOnce).toBe(true)
})

it('uses a custom theme for all child components if one is provided', () => {
  const customTheme = {
    colors: {
      action: 'green',
      actionHover: 'darkgreen',
      primaryBackground: 'white',
      primaryForeground: 'green',
      secondaryBackground: 'green',
      secondaryForeground: 'white',
      secondaryForegroundFocus: 'gray',
    },
  }

  const tree = renderer
    .create(
      <StyleRoot>
        <div>
          <TextField
            id="test_id"
            name="test"
            type="email"
            floatingLabelText="Email"
            hintText="Enter your email address"
            onChange={() => {}}
            snacksTheme={customTheme}
          />
        </div>
      </StyleRoot>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
