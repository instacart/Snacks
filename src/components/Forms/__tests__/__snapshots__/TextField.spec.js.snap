// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders TextField correctly 1`] = `
<div
  data-radium={true}
>
  <div>
    <div
      data-radium={true}
      style={
        Object {
          "cursor": "auto",
          "display": "inline-block",
          "position": "relative",
          "width": "343px",
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "borderRadius": "4px",
            "position": "relative",
          }
        }
      >
        <label
          data-radium={true}
          htmlFor="test_id"
          style={
            Object {
              "color": "#757575",
              "cursor": "text",
              "fontSize": "16px",
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "pointerEvents": "auto",
              "position": "absolute",
              "transform": "scale(1) translate(8px, 17px)",
              "transformOrigin": "left top",
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "userSelect": "none",
              "zIndex": 1,
            }
          }
        >
          Email
        </label>
        <div
          data-radium={true}
          style={
            Object {
              "color": "#BDBDBD",
              "cursor": "inherit",
              "fontSize": "16px",
              "opacity": 0,
              "pointerEvents": "none",
              "position": "absolute",
              "transform": "scale(1) translate(9px, 26px)",
              "transformOrigin": "left top",
              "transition": "opacity 250ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "zIndex": 1,
            }
          }
        >
          Enter your email address
        </div>
        <input
          aria-describedby={null}
          aria-invalid={false}
          aria-required={undefined}
          autoComplete="on"
          data-radium={true}
          defaultValue={null}
          disabled={false}
          id="test_id"
          name="test"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder=""
          style={
            Object {
              "WebkitOpacity": 1,
              "WebkitTapHighlightColor": "rgba(0,0,0,0)",
              "backgroundColor": "#FFF",
              "border": "solid 1px #BDBDBD",
              "borderRadius": "4px",
              "boxSizing": "border-box",
              "color": "#424242",
              "fontSize": "16px",
              "height": "56px",
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "outline": "none",
              "paddingBottom": 8,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": "25px",
              "position": "relative",
              "width": "100%",
            }
          }
          type="email"
          value={undefined}
        />
      </div>
      <div
        aria-atomic={true}
        aria-live="assertive"
        data-radium={true}
        id="error_test_id"
        style={
          Object {
            "color": "#B30029",
            "display": "none",
            "fontSize": "12px",
            "marginLeft": "1px",
            "marginTop": "2px",
            "opacity": "0",
            "pointerEvents": "none",
            "textAlign": "left",
            "transition": "opacity 250ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          }
        }
      />
    </div>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`uses a custom theme for all child components if one is provided 1`] = `
<div
  data-radium={true}
>
  <div>
    <div
      data-radium={true}
      style={
        Object {
          "cursor": "auto",
          "display": "inline-block",
          "position": "relative",
          "width": "343px",
        }
      }
    >
      <div
        data-radium={true}
        style={
          Object {
            "borderRadius": "4px",
            "position": "relative",
          }
        }
      >
        <label
          data-radium={true}
          htmlFor="test_id"
          style={
            Object {
              "color": "#757575",
              "cursor": "text",
              "fontSize": "16px",
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "pointerEvents": "auto",
              "position": "absolute",
              "transform": "scale(1) translate(8px, 17px)",
              "transformOrigin": "left top",
              "transition": "all 450ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "userSelect": "none",
              "zIndex": 1,
            }
          }
        >
          Email
        </label>
        <div
          data-radium={true}
          style={
            Object {
              "color": "#BDBDBD",
              "cursor": "inherit",
              "fontSize": "16px",
              "opacity": 0,
              "pointerEvents": "none",
              "position": "absolute",
              "transform": "scale(1) translate(9px, 26px)",
              "transformOrigin": "left top",
              "transition": "opacity 250ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
              "zIndex": 1,
            }
          }
        >
          Enter your email address
        </div>
        <input
          aria-describedby={null}
          aria-invalid={false}
          aria-required={undefined}
          autoComplete="on"
          data-radium={true}
          defaultValue={null}
          disabled={false}
          id="test_id"
          name="test"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          placeholder=""
          style={
            Object {
              "WebkitOpacity": 1,
              "WebkitTapHighlightColor": "rgba(0,0,0,0)",
              "backgroundColor": "#FFF",
              "border": "solid 1px #BDBDBD",
              "borderRadius": "4px",
              "boxSizing": "border-box",
              "color": "#424242",
              "fontSize": "16px",
              "height": "56px",
              "marginBottom": 0,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 0,
              "outline": "none",
              "paddingBottom": 8,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": "25px",
              "position": "relative",
              "width": "100%",
            }
          }
          type="email"
          value={undefined}
        />
      </div>
      <div
        aria-atomic={true}
        aria-live="assertive"
        data-radium={true}
        id="error_test_id"
        style={
          Object {
            "color": "#B30029",
            "display": "none",
            "fontSize": "12px",
            "marginLeft": "1px",
            "marginTop": "2px",
            "opacity": "0",
            "pointerEvents": "none",
            "textAlign": "left",
            "transition": "opacity 250ms cubic-bezier(0.23, 1, 0.32, 1) 0ms",
          }
        }
      />
    </div>
  </div>
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;
